test/junit/scala/tools/nsc/backend/jvm/opt/InlinerTest.scala

```diff
@@ -1443,17 +1443,11 @@ class InlinerTest extends BytecodeTesting {
     val c = compileClass(code)
 
     // box-unbox will clean it up
-    try assertSameSummary(getMethod(c, "t"), List(
+    assertSameSummary(getMethod(c, "t"), List(
       ALOAD, "$anonfun$t$1", IFEQ /*A*/,
-      "$anonfun$t$2", IRETURN,
-      -1 /*A*/, "$anonfun$t$3", IRETURN))
-    catch { case e: AssertionError =>
-      try assertSameSummary(getMethod(c, "t"), List( // this is the new behaviour, after restarr'ing
-        ALOAD, "debug", IFEQ /*A*/,
-        "$anonfun$t$2", IRETURN,
-        -1 /*A*/, "$anonfun$t$3", IRETURN))
-      catch { case _: AssertionError => throw e }
-    }
+      "$anonfun$t$2", ISTORE, GOTO /*B*/,
+      -1 /*A*/, "$anonfun$t$3", ISTORE,
+      -1 /*B*/, ILOAD, IRETURN))
```

Comment:
@lrytz Do you have any idea why the optimizer generates *worse* code after these changes than before? With the changes, with or without optimizer, the bytecode summary generated for `ifelse` is
```
ALOAD, "apply$mcZ$sp", IFEQ /*8*/, ALOAD, "apply", ARETURN, -1 /*8*/, ALOAD, "apply", ARETURN
```
So why does the generated code for `t` regresses to having an `ISTORE`/`GOTO`/`ILOAD`?

Or perhaps a better question: how did the optimizer have the direct `IRETURN`s *before* the changes?

Do you have any pointer in the optimizer to help me figure this out?

Reply:
It seems the difference in the generated code is that the `store` is pushed into the two branches. before there were jumps to the single store instruction.

`sc12 Test.scala -opt:l:inline '-opt-inline-from:**' -Yopt-trace C.t`


```diff
@@ -65,27 +65,27 @@
     POP
     INVOKESTATIC C.$anonfun$t$2 ()I
     INVOKESTATIC scala/runtime/BoxesRunTime.boxToInteger (I)Ljava/lang/Integer;
+    ASTORE 5
     GOTO L9
    L7
     ALOAD 4
     POP
     INVOKESTATIC C.$anonfun$t$3 ()I
     INVOKESTATIC scala/runtime/BoxesRunTime.boxToInteger (I)Ljava/lang/Integer;
-   L9
     ASTORE 5
-    GOTO L10
-   L11
+    GOTO L9
    L10
+   L9
     ALOAD 5
-    LINENUMBER 4 L10
+    LINENUMBER 4 L9
     INVOKESTATIC scala/runtime/BoxesRunTime.unboxToInt (Ljava/lang/Object;)I
     IRETURN
-   L12
+   L11
```

So the store-load-pair removal no longer triggers

https://github.com/scala/scala/blob/7dbc139b4a113c18e99e8720b985a249fc9b181f/src/compiler/scala/tools/nsc/backend/jvm/opt/CopyProp.scala#L476-L500

Reply:
Thanks. That sounds plausible indeed! I'll see if I can do something to recover from that.

Reply:
I pushed a commit in the inliner that addresses this regression.